TSOClient/tso.client/UI/Controls/UIMapWaypoint.cs

```diff
@@ -0,0 +1,323 @@
+                    Position = new Vector2(other.Position.X - 10 - SizeW, animNewLoc.Y); // try left
+                    if (Position.X < 10)
+                        Position = new Vector2(other.Position.X + 10 + SizeW, animNewLoc.Y); // fallback on right
+                    completion = 1; // make sure the waypoint cannot animate back to its desired position!
+                }
+            }
+        }
+
+        private void ensureOnScreen()
+        {
+            var gamescreen = ((CoreGameScreen)GameFacade.Screens.CurrentUIScreen);
+            if (animNewLoc.X < 0) {
+                animNewLoc.X = 10;
+                completion = 0;
+            }
+            if (animNewLoc.X + SizeW > gamescreen.GetBounds().Width)
+            {
+                animNewLoc.X = gamescreen.GetBounds().Width - SizeW - 10;
+                completion = 0;
+            }
+            if (animNewLoc.Y < 0)
+            { 
+                animNewLoc.Y = 10;
+                completion = 0;
+            }
+            if (animNewLoc.Y + SizeH > gamescreen.GetBounds().Height)
+            {
+                animNewLoc.Y = gamescreen.GetBounds().Height - SizeH - 10;
+                completion = 0;
+            }
+        }
+
+        private Vector2 getDesiredLocation()
+        {
+            var xp = (int)_LotId >> 16;
+            var yp = (int)_LotId & 0xFFFF;
+
+            var terrain = ((CoreGameScreen)GameFacade.Screens.CurrentUIScreen).CityRenderer;
+            return ((terrain.GetFar2DFromTile(xp, yp) + terrain.GetFar2DFromTile(xp + 1, yp + 1)) / 2) - new Vector2(SizeW / 2,150 + SizeH);
+        }
+
+        private void reposition()
+        {
+            if (!Is3D)
+            {
+                var pos = getDesiredLocation();
+                if (double.IsNaN(pos.X) || double.IsNaN(pos.Y))
+                {
+                    attemptRedraw = true;
+                    return;
+                }
+                attemptRedraw = false;
+                Position = pos - new Vector2(SizeW/2, 150 - SizeH); // this is on purpose: makes the marker sprout up to alert the user
+            }
+        }       
+
+        protected void OnButtonClick()
+        {
+            FindController<CoreGameScreenController>().ShowLotPage(_LotId);
+        }       
+
+        public override Rectangle GetBounds()
+        {
+            return new Rectangle((int)Position.X,(int)Position.Y, SizeW, SizeH);
+        }  
+
+        public void update3DZindex()
+        {
+            var xp = (int)_LotId >> 16;
+            var yp = (int)_LotId & 0xFFFF;
+
+            var terrain = ((CoreGameScreen)GameFacade.Screens.CurrentUIScreen).CityRenderer;
+            var pos = terrain.transformSpr4(new Vector3(xp+0.5f, terrain.InterpElevationAt(new Vector2(xp + 0.5f, yp + 0.5f))+2f, yp+0.5f));
+
+            ZOrder = pos.Z;
+        }
+
+        public override void Update(UpdateState state)
+        {                        
+            base.Update(state);
```

Comment:
Note: if you base.Update() after changing position, you may be able to remove the one frame latency. That's just a theory, though...